File: benchmarks/list_access/python/list_access.py
#!/usr/bin/env python3
import os
import time

LENGTH = 65536
PASSES = 128
OPERATIONS = LENGTH * PASSES
VALUES = [(index * 17 + 3) % 1009 for index in range(LENGTH)]
PARTICLES = [
    (index % 97, index % 89, index % 83, index % 79,
     index % 73, index % 71, index % 67, index % 61)
    for index in range(LENGTH)
]


def benchmark(name, body):
    start = time.perf_counter_ns()
    checksum = body()
    elapsed = time.perf_counter_ns() - start
    print(f"RESULT,python,{name},{elapsed},{checksum}")


def int_sequential_native():
    checksum = 0
    for _pass in range(PASSES):
        for index in range(LENGTH):
            checksum += VALUES[index]
    return checksum


def int_sequential_checked():
    checksum = 0
    for _pass in range(PASSES):
        for index in range(LENGTH):
            value = VALUES[index] if 0 <= index < len(VALUES) else None
            if value is not None:
                checksum += value
    return checksum


def int_collection():
    checksum = 0
    for _pass in range(PASSES):
        for value in VALUES:
            checksum += value
    return checksum


def int_constant_checked():
    checksum = 0
    index = int(os.environ.get("BENCHMARK_INDEX", "137"))
    for _iteration in range(OPERATIONS):
        value = VALUES[index] if 0 <= index < len(VALUES) else None
        if value is not None:
            checksum += value
    return checksum


def int_strided_checked():
    checksum = 0
    index = 0
    for _iteration in range(OPERATIONS):
        value = VALUES[index] if 0 <= index < len(VALUES) else None
        if value is not None:
            checksum += value
        index = (index + 4) % LENGTH
    return checksum


def int_random_checked():
    checksum = 0
    for iteration in range(OPERATIONS):
        index = (iteration * 48271 + 17) % LENGTH
        value = VALUES[index] if 0 <= index < len(VALUES) else None
        if value is not None:
            checksum += value
    return checksum


def int_mostly_valid_checked():
    checksum = 0
    for iteration in range(OPERATIONS):
        index = LENGTH if iteration % 1000 == 0 else (iteration * 48271 + 17) % LENGTH
        value = VALUES[index] if 0 <= index < len(VALUES) else None
        if value is not None:
            checksum += value
    return checksum


def int_mixed_invalid_checked():
    checksum = 0
    for iteration in range(OPERATIONS):
        index = -1 if iteration % 4 == 0 else (iteration * 48271 + 17) % LENGTH
        value = VALUES[index] if 0 <= index < len(VALUES) else None
        if value is not None:
            checksum += value
    return checksum


def struct_sequential_ref():
    checksum = 0
    for _pass in range(PASSES):
        for index in range(LENGTH):
            particle = PARTICLES[index] if 0 <= index < len(PARTICLES) else None
            if particle is not None:
                checksum += particle[0] + particle[5] + particle[6]
    return checksum


def struct_collection_ref():
    checksum = 0
    for _pass in range(PASSES):
        for particle in PARTICLES:
            checksum += particle[0] + particle[5] + particle[6]
    return checksum


def struct_random_ref():
    checksum = 0
    for iteration in range(OPERATIONS):
        index = (iteration * 48271 + 17) % LENGTH
        particle = PARTICLES[index] if 0 <= index < len(PARTICLES) else None
        if particle is not None:
            checksum += particle[0] + particle[5] + particle[6]
    return checksum


SCENARIOS = [
    ("int_sequential_native", int_sequential_native),
    ("int_sequential_checked", int_sequential_checked),
    ("int_collection", int_collection),
    ("int_constant_checked", int_constant_checked),
    ("int_strided_checked", int_strided_checked),
    ("int_random_checked", int_random_checked),
    ("int_mostly_valid_checked", int_mostly_valid_checked),
    ("int_mixed_invalid_checked", int_mixed_invalid_checked),
    ("struct_sequential_ref", struct_sequential_ref),
    ("struct_collection_ref", struct_collection_ref),
    ("struct_random_ref", struct_random_ref),
]

for _repetition in range(9):
    for scenario_name, scenario in SCENARIOS:
        benchmark(scenario_name, scenario)

File: benchmarks/list_access/javascript/list_access.js
"use strict";

const length = 65536;
const passes = 128;
const operations = length * passes;
const values = Array.from({length}, (_, index) => (index * 17 + 3) % 1009);
const particles = Array.from({length}, (_, index) => ({
  px: index % 97, py: index % 89, pz: index % 83, vx: index % 79,
  vy: index % 73, vz: index % 71, mass: index % 67, flags: index % 61,
}));

function benchmark(name, body) {
  const start = process.hrtime.bigint();
  const checksum = body();
  const elapsed = process.hrtime.bigint() - start;
  console.log(`RESULT,javascript,${name},${elapsed},${checksum}`);
}

function runSuite() {
  benchmark("int_sequential_native", () => {
    let checksum = 0;
    for (let pass = 0; pass < passes; pass++)
      for (let index = 0; index < length; index++) checksum += values[index];
    return checksum;
  });
  benchmark("int_sequential_checked", () => {
    let checksum = 0;
    for (let pass = 0; pass < passes; pass++) {
      for (let index = 0; index < length; index++) {
        const value = index >= 0 && index < values.length ? values[index] : undefined;
        if (value !== undefined) checksum += value;
      }
    }
    return checksum;
  });
  benchmark("int_collection", () => {
    let checksum = 0;
    for (let pass = 0; pass < passes; pass++)
      for (const value of values) checksum += value;
    return checksum;
  });
  benchmark("int_constant_checked", () => {
    let checksum = 0;
    const index = Number(process.env.BENCHMARK_INDEX || 137);
    for (let iteration = 0; iteration < operations; iteration++) {
      const value = index >= 0 && index < values.length ? values[index] : undefined;
      if (value !== undefined) checksum += value;
    }
    return checksum;
  });
  benchmark("int_strided_checked", () => {
    let checksum = 0;
    let index = 0;
    for (let iteration = 0; iteration < operations; iteration++) {
      const value = index >= 0 && index < values.length ? values[index] : undefined;
      if (value !== undefined) checksum += value;
      index = (index + 4) % length;
    }
    return checksum;
  });
  benchmark("int_random_checked", () => {
    let checksum = 0;
    for (let iteration = 0; iteration < operations; iteration++) {
      const index = (iteration * 48271 + 17) % length;
      const value = index >= 0 && index < values.length ? values[index] : undefined;
      if (value !== undefined) checksum += value;
    }
    return checksum;
  });
  benchmark("int_mostly_valid_checked", () => {
    let checksum = 0;
    for (let iteration = 0; iteration < operations; iteration++) {
      const index = iteration % 1000 === 0 ? length : (iteration * 48271 + 17) % length;
      const value = index >= 0 && index < values.length ? values[index] : undefined;
      if (value !== undefined) checksum += value;
    }
    return checksum;
  });
  benchmark("int_mixed_invalid_checked", () => {
    let checksum = 0;
    for (let iteration = 0; iteration < operations; iteration++) {
      const index = iteration % 4 === 0 ? -1 : (iteration * 48271 + 17) % length;
      const value = index >= 0 && index < values.length ? values[index] : undefined;
      if (value !== undefined) checksum += value;
    }
    return checksum;
  });
  benchmark("struct_sequential_ref", () => {
    let checksum = 0;
    for (let pass = 0; pass < passes; pass++) {
      for (let index = 0; index < length; index++) {
        const particle = index >= 0 && index < particles.length ? particles[index] : undefined;
        if (particle !== undefined) checksum += particle.px + particle.vz + particle.mass;
      }
    }
    return checksum;
  });
  benchmark("struct_collection_ref", () => {
    let checksum = 0;
    for (let pass = 0; pass < passes; pass++)
      for (const particle of particles) checksum += particle.px + particle.vz + particle.mass;
    return checksum;
  });
  benchmark("struct_random_ref", () => {
    let checksum = 0;
    for (let iteration = 0; iteration < operations; iteration++) {
      const index = (iteration * 48271 + 17) % length;
      const particle = index >= 0 && index < particles.length ? particles[index] : undefined;
      if (particle !== undefined) checksum += particle.px + particle.vz + particle.mass;
    }
    return checksum;
  });
}

for (let repetition = 0; repetition < 9; repetition++) runSuite();
